src/field_theory/primitive_element.lean

```diff
@@ -0,0 +1,260 @@
+/-
+Copyright (c) 2020 Thomas Browning and Patrick Lutz. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Thomas Browning and Patrick Lutz
+-/
+
+import field_theory.adjoin
+import field_theory.separable
+
+/-!
+# Primitive Element Theorem
+
+In this file we prove the primitive element theorem.
+
+## Main results
+
+- `primitive_element`: a finite separable extension has a primitive element.
+
+-/
+
+noncomputable theory
+open_locale classical
+
+/- Proof of the primitive element theorem. -/
+
+open finite_dimensional
+
+namespace field
+
+section primitive_element_finite
+variables (F : Type*) [field F] {E : Type*} [field E] [algebra F E]
+
+/-! ### Primitive element theorem for finite fields -/
+
+/-- Primitive element theorem assuming E is finite. -/
+lemma primitive_element_fin_aux [fintype E] : ∃ α : E, F⟮α⟯ = ⊤ :=
+begin
+  obtain ⟨α, hα⟩ := is_cyclic.exists_generator (units E),
+  use α,
+  ext,
+  rw iff_true_right algebra.mem_top,
+  by_cases hx : x = 0,
+  { rw hx,
+    exact F⟮α.val⟯.zero_mem, },
+  { obtain ⟨n, hn⟩ := set.mem_range.mp (hα (units.mk0 x hx)),
+    rw (show x = α^n, by { norm_cast, simp * }),
+    exact @is_subfield.pow_mem E _ α.val n F⟮α.val⟯ _ (field.mem_adjoin_simple_self F α.val), },
+end
+
+/-- Primitive element theorem for finite dimensional extension of a finite field. -/
+theorem primitive_element_fin [fintype F] [hfd : finite_dimensional F E] :
```

Comment:
Isn't this name confusing because the primitive element theorem always assumes the field extension is finite dimensional? The thing that makes this theorem statement different from later ones in the file is that it assumes `F` is finite.

Reply:
Good point, maybe the first one can be `of_fintype_top` and this one `of_fintype_bottom`?

Reply:
How do you feel about calling it `primitive_element_of_fintype_bot`?